frontend/src/data/MetricQuery.ts

```diff
@@ -48,6 +48,10 @@ export class MetricQueryResponse {
       this.error = input as Error;
       this.data = [];
       this.invalidValues = {};
+    } else if (input.length <= 0) {
+      this.error = new ExpectedError("No rows returned");
+      this.data = [];
+      this.invalidValues = {};
```

Comment:
I'm not sure if failure is the best term. To me, it implies success was potentially possible. But the situations that would result in this lack of data are: 
1. unsupported breakdown 2. dataset rows are empty (for example, user requests a state for which the dataset doesn't have data)
as opposed to situations situations that throw instead of returning MetricQueryResponse:
1. invalid dataset id 2. server failures

Maybe "unavailable"or "invalid" or "bad request" would be more accurate than "error" or "failure"?

I've changed "isError" to "dataIsUnavailable" and the "showErrorMessage" to "shouldShowMissingDataMessage" since the actual text of the alert we're showing in this situation does say "data is missing".

What do you think?

I'm also not sure we need isFailure as a constructor input... couldn't this be derived based on absence/presence of failureMessage/dataUnavailableMessage ?

Also, do you have a specific opposition to having the input just be either message or the data?

Reply:
Your reasoning makes sense. I'm not specifically attached to "failure" terminology, I just don't want to use "error" and failure was the best alternative I could think of because I thought others were potentially too specific. But data unavailable works too, I think it covers all the use cases.

re: input as (Row[] | string) and deriving failure based on presence of the message, I can see pros/cons to both approaches. The only downsides to doing it the way you suggested are that it doesn't scale as well if you need more granular/structured information about the failure, and it's a little less explicit with what's going on. But it works for the current use case, and it's not hard to change later if we find ourselves needing more structured information. I'd suggest adding a comment though indicating that when the data provided is a string, it is treated as an error. I think it's kind of subtle. I also still think a factory method that only accepts a string and is named clearly, like `createMissingDataResponse(message: string)` helps improve readability.

Reply:
i added createMissingDataResponse and made all names refer to "missing data" per slack chat